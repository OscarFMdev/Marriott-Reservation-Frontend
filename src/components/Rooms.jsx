import { useSelector } from 'react-redux';
import { ProgressSpinner } from 'primereact/progressspinner';
import Carousel from './Carousel';
import stl from './componentsCss/Carousel.module.css';

const Rooms = () => {
  const rooms = useSelector((state) => state.rooms);

  if (rooms.status === 'loading') {
    return <ProgressSpinner className="progressSpinner" />;
  }
  return (
    <div className={stl.room_container}>
      <h2 className={stl.title}>Available Rooms</h2>
      {rooms.rooms.length === 0 && <h1>No Rooms</h1>}
      <Carousel />
    </div>
  );
};

export default Rooms;
