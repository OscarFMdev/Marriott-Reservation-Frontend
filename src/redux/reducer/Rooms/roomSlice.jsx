import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const BASEURL = 'https://marriott-reservation.onrender.com/api/v1';

const initialState = {
  rooms: [],
  status: '',
  message: '',
  error: '',
};

export const fetchRooms = createAsyncThunk('rooms/fetchRooms', async () => {
  const response = await fetch(`${BASEURL}/rooms`);
  const rooms = await response.json();
  return rooms;
});

export const addRoom = createAsyncThunk('rooms/addRoom', async (room) => {
  const response = await fetch(`${BASEURL}/rooms`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: localStorage.getItem('token'),
    },
    body: JSON.stringify({ room }),
  });

  const { status: code } = response;
  if (code === 201) {
    const { data, message } = await response.json();
    return {
      status: '',
      message,
      data,
    };
  }

  if (code === 401) {
    const { message } = await response.json();
    return {
      status: 'error',
      message,
    };
  }

  if (code === 422) {
    const { message } = await response.json();
    return {
      status: 'error',
      message,
    };
  }

  return null;
});
export const deleteRoom = createAsyncThunk('rooms/deleteRoom', async (id) => {
  const response = await fetch(`${BASEURL}/rooms/${id}`, {
    method: 'DELETE',
    headers: {
      Authorization: localStorage.getItem('token'),
    },
  });
  const deletedRoom = await response.json();
  return deletedRoom;
});

const roomsSlice = createSlice({
  name: 'rooms',
  initialState,
  reducers: {
    NewRoom: (state, action) => ({
      ...state,
      rooms: [...state.rooms, action.payload],
    }),
    PopRoom: (state, action) => ({
      ...state,
      rooms: state.rooms.filter((room) => room.id !== action.payload),
    }),
    setMessageEmpty: (state, action) => ({
      ...state,
      message: action.payload,
    }),
  },
  extraReducers: (builder) => {
    builder
      .addCase(addRoom.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(addRoom.fulfilled, (state, action) => ({
        ...state,
        rooms: [
          ...(action.payload.status === 201 ? action.payload.data : []),
          ...state.rooms,
        ],
        message: action.payload.message,
        status: action.payload.status,
      }))
      .addCase(addRoom.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }))
      .addCase(fetchRooms.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(fetchRooms.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        rooms: action.payload,
      }))
      .addCase(fetchRooms.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }))
      .addCase(deleteRoom.pending, (state) => ({
        ...state,
        status: 'loading',
      }))
      .addCase(deleteRoom.fulfilled, (state, action) => ({
        ...state,
        status: 'success',
        rooms: state.rooms.filter((room) => room.id !== action.payload),
        message: action.payload.message,
      }))
      .addCase(deleteRoom.rejected, (state, action) => ({
        ...state,
        status: 'error',
        message: action.payload.message,
      }));
  },
});

export default roomsSlice.reducer;
export const { NewRoom, PopRoom, setMessageEmpty } = roomsSlice.actions;
