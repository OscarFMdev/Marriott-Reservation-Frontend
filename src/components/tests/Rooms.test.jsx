import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Rooms from '../Rooms';
import store from '../../redux/configureStore';

describe('It displays the Roooms title', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <Router>
          <Rooms />
        </Router>
      </Provider>,
    );
  });
  test('List of Rooms', () => {
    expect(screen.getByText('Available Rooms')).toBeDefined();
  });
});

test('Rooms renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <Rooms />
      </Router>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
