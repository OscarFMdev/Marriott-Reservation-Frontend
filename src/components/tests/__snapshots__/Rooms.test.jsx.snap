// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Rooms renders correctly 1`] = `
<div
  className="_room_container_428451"
>
  <h2
    className="_title_428451"
  >
    Available Rooms
  </h2>
  <h1>
    No Rooms
  </h1>
</div>
`;
